lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -31,9 +31,40 @@
 from ansible.module_utils.six import string_types
 from ansible.utils.version import SemanticVersion
 
+from collections.abc import Set
 from resolvelib import AbstractProvider
 
 
+class PinnedCandidateRequests(Set):
+    """Custom set class to store Candidate objects. Excludes the 'signatures' attribute when determining if a Candidate instance is in the set."""
+    CANDIDATE_ATTRS = ('fqcn', 'ver', 'src', 'type')
+
+    def __init__(self, candidates):
+        self._candidates = set(candidates)
+
+    def __iter__(self):
+        return iter(self._candidates)
+
+    def __contains__(self, value):
+        if not isinstance(value, Candidate):
+            raise ValueError(
+                "Expected a Candidate object but got {value} ({type(value)})"
```

Comment:
Forgotten `f`-prefix?
```suggestion
                f"Expected a Candidate object but got {value} ({type(value)})"
```

Reply:
Alternatively, could use repr:
```suggestion
                f"Expected a Candidate object but got {value!r}"
```

Reply:
I like the repr suggestion. Fixed.